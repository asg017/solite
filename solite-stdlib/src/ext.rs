use arboard::Clipboard;
use sqlite_loadable::prelude::*;
use sqlite_loadable::{api, define_scalar_function, FunctionFlags, Result};

pub fn solite_stdlib_version(
    context: *mut sqlite3_context,
    _values: &[*mut sqlite3_value],
) -> Result<()> {
    api::result_text(context, format!("v{}", env!("CARGO_PKG_VERSION")))?;
    Ok(())
}

pub fn clipboard_set(context: *mut sqlite3_context, values: &[*mut sqlite3_value]) -> Result<()> {
    let contents = api::value_text(&values[0])?;
    let mut cb = Clipboard::new().unwrap();
    cb.set_text(contents).unwrap();
    api::result_bool(context, true);
    Ok(())
}

#[sqlite_entrypoint]
pub fn sqlite3_solite_stdlib_init(db: *mut sqlite3) -> Result<()> {
    define_scalar_function(
        db,
        "solite_stdlib_version",
        0,
        solite_stdlib_version,
        FunctionFlags::UTF8 | FunctionFlags::DETERMINISTIC,
    )?;
    define_scalar_function(db, "clipboard_set", 1, clipboard_set, FunctionFlags::UTF8)?;
    Ok(())
}
