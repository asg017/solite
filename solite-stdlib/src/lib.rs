mod ext;

use crate::ext::sqlite3_solite_stdlib_init;
use sqlite_http::sqlite3_http_init;
use sqlite_loadable::ext::{sqlite3, sqlite3_api_routines};
use sqlite_path::sqlite3_path_init;
use sqlite_regex::sqlite3_regex_init;
use sqlite_ulid::sqlite3_ulid_init;
use sqlite_url::sqlite3_url_init;
use sqlite_vec::sqlite3_vec_init;
use std::ffi::{c_char, c_uint};
//use sqlite_fastrand::sqlite3_fastrand_init;
use sqlite_xsv::sqlite3_xsv_init;

include!(concat!(env!("OUT_DIR"), "/builtins.rs"));

#[link(name = "base64")]
extern "C" {
    fn sqlite3_base_init();
}
//#[link(name = "base85")]
//extern "C" {
//    fn sqlite3_base85_init();
//}
#[link(name = "decimal")]
extern "C" {
    fn sqlite3_decimal_init();
}
#[link(name = "ieee754")]
extern "C" {
    fn sqlite3_ieee_init();
}
#[link(name = "fileio")]
extern "C" {
    fn sqlite3_fileio_init();
}
#[link(name = "uint")]
extern "C" {
    fn sqlite3_uint_init();
}
#[link(name = "series")]
extern "C" {
    fn sqlite3_series_init();
}
#[link(name = "sha1")]
extern "C" {
    fn sqlite3_sha_init();
}
#[link(name = "shathree")]
extern "C" {
    fn sqlite3_shathree_init();
}
#[link(name = "spellfix")]
extern "C" {
    fn sqlite3_spellfix_init();
}

/*
#[link(name = "sqlar")]
extern "C" {
    fn sqlite3_sqlar_init();
}
#[link(name = "compress")]
extern "C" {
    fn sqlite3_compress_init();
}
*/
#[link(name = "uuid")]
extern "C" {
    fn sqlite3_uuid_init();
}

/*
#[link(name = "zipfile")]
extern "C" {
    fn sqlite3_zipfile_init();
}
*/

#[link(name = "usleep")]
extern "C" {
    fn sqlite3_usleep_init();
}

#[link(name = "completion")]
extern "C" {
    fn sqlite3_completion_init();
}

unsafe fn init_arg0(
    func: unsafe extern "C" fn(),
    db: *mut sqlite3,
    pz_err_msg: *mut *mut c_char,
    p_api: *mut sqlite3_api_routines,
) {
    let x: unsafe extern "C" fn(
        db: *mut sqlite3,
        pz_err_msg: *mut *mut c_char,
        p_api: *mut sqlite3_api_routines,
    ) -> c_uint = std::mem::transmute(func as *const ());
    x(db, pz_err_msg, p_api);
}
/// # Safety
/// lol
#[no_mangle]
pub unsafe extern "C" fn solite_stdlib_init(
    db: *mut sqlite3,
    pz_err_msg: *mut *mut c_char,
    p_api: *mut sqlite3_api_routines,
) -> c_uint {
    sqlite3_ulid_init(db, pz_err_msg, p_api);
    sqlite3_regex_init(db, pz_err_msg, p_api);
    sqlite3_http_init(db, pz_err_msg, p_api);
    sqlite3_path_init(db, pz_err_msg, p_api);
    sqlite3_url_init(db, pz_err_msg, p_api);
    sqlite3_xsv_init(db, pz_err_msg, p_api);

    sqlite3_solite_stdlib_init(db, pz_err_msg, p_api);

    init_arg0(sqlite3_base_init, db, pz_err_msg, p_api);

    //sqlite3_fastrand_init(db, pz_err_msg, p_api);
    init_arg0(sqlite3_vec_init, db, pz_err_msg, p_api);
    init_arg0(sqlite3_completion_init, db, pz_err_msg, p_api);
    init_arg0(sqlite3_decimal_init, db, pz_err_msg, p_api);
    init_arg0(sqlite3_ieee_init, db, pz_err_msg, p_api);
    init_arg0(sqlite3_fileio_init, db, pz_err_msg, p_api);
    init_arg0(sqlite3_uint_init, db, pz_err_msg, p_api);
    init_arg0(sqlite3_series_init, db, pz_err_msg, p_api);
    init_arg0(sqlite3_sha_init, db, pz_err_msg, p_api);
    init_arg0(sqlite3_shathree_init, db, pz_err_msg, p_api);
    init_arg0(sqlite3_spellfix_init, db, pz_err_msg, p_api);
    //init_arg0(sqlite3_sqlar_init, db, pz_err_msg, p_api);
    //init_arg0(sqlite3_compress_init, db, pz_err_msg, p_api);
    init_arg0(sqlite3_uuid_init, db, pz_err_msg, p_api);
    init_arg0(sqlite3_usleep_init, db, pz_err_msg, p_api);
    //init_arg0(sqlite3_zipfile_init, db, pz_err_msg, p_api);
    //init_arg0(sqlite3_base85_init, db, pz_err_msg, p_api);
    0
}

// TODO different modes
// safe mode: HTTP, usleep, zipfile, fileio

#[cfg(test)]
mod tests {
    use super::*;
    use rusqlite::Connection;

    fn functions_of(db: Connection) -> Vec<String> {
        db.prepare("select distinct name from pragma_function_list order by 1")
            .unwrap()
            .query_map([], |row| Ok(row.get::<usize, String>(0).unwrap()))
            .unwrap()
            .collect::<Result<Vec<_>, _>>()
            .unwrap()
    }
    #[test]
    fn stdlib_basic() {
        let db = Connection::open_in_memory().unwrap();

        assert_eq!(
            db.query_row("select sqlite_version();", [], |r| r
                .get::<usize, String>(0))
                .unwrap(),
            "3.49.1"
        );

        let base_functions: Vec<String> = functions_of(db);
        insta::assert_yaml_snapshot!(base_functions);

        let db = Connection::open_in_memory().unwrap();
        unsafe {
            solite_stdlib_init(db.handle(), std::ptr::null_mut(), std::ptr::null_mut());
        }
        /*assert!(db
        .query_row("select ulid_version();", [], |r| r.get::<usize, String>(0))
        .unwrap()
        .starts_with('v'));*/
        let solite_functions: Vec<String> = functions_of(db);
        let solite_only_functions: Vec<String> = solite_functions
            .iter()
            .filter(|v| !base_functions.contains(v))
            .map(|v| v.to_string())
            .collect::<Vec<String>>();
        insta::assert_yaml_snapshot!(solite_only_functions);

        insta::assert_yaml_snapshot!(BUILTIN_FUNCTIONS);
    }
}
