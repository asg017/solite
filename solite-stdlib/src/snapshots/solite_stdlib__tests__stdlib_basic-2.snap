---
source: solite-stdlib/src/lib.rs
expression: solite_only_functions
snapshot_kind: text
---
- base64
- clipboard_set
- compress
- csv_at
- csv_line_number
- csv_path
- decimal
- decimal_add
- decimal_cmp
- decimal_exp
- decimal_mul
- decimal_pow2
- decimal_sub
- decimal_sum
- editdist3
- http_debug
- http_get_body
- http_request
- http_version
- ieee754
- ieee754_exponent
- ieee754_from_blob
- ieee754_inc
- ieee754_mantissa
- ieee754_to_blob
- lines_debug
- lines_version
- lsmode
- path_at
- path_debug
- path_version
- readfile
- regex
- regex_capture
- regex_debug
- regex_find
- regex_find_at
- regex_print
- regex_replace
- regex_replace_all
- regex_valid
- regex_version
- regexp
- regexset
- regexset_is_match
- regexset_print
- sha1
- sha1_query
- sha1b
- sha3
- sha3_agg
- sha3_query
- solite_stdlib_version
- spellfix1_editdist
- spellfix1_phonehash
- spellfix1_scriptcode
- spellfix1_translit
- sqlar_compress
- sqlar_uncompress
- tg_contains
- tg_coveredby
- tg_covers
- tg_debug
- tg_disjoint
- tg_extra_json
- tg_geom
- tg_group_feature_collection_geojson
- tg_group_geometry_collection
- tg_group_multipoint
- tg_intersects
- tg_multipoint
- tg_point
- tg_to_geojson
- tg_to_wkb
- tg_to_wkt
- tg_touches
- tg_type
- tg_valid_geojson
- tg_valid_wkb
- tg_valid_wkt
- tg_version
- tg_within
- tsv_at
- tsv_line_number
- tsv_path
- ulid
- ulid_bytes
- ulid_datetime
- ulid_debug
- ulid_version
- ulid_with_datetime
- ulid_with_prefix
- uncompress
- url_debug
- url_escape
- url_fragment
- url_host
- url_path
- url_port
- url_query
- url_scheme
- url_valid
- url_version
- usleep
- uuid
- uuid_blob
- uuid_str
- vec_add
- vec_bit
- vec_debug
- vec_distance_cosine
- vec_distance_hamming
- vec_distance_l1
- vec_distance_l2
- vec_f32
- vec_int8
- vec_length
- vec_normalize
- vec_quantize_binary
- vec_quantize_int8
- vec_slice
- vec_sub
- vec_to_json
- vec_type
- vec_version
- writefile
- xsv_at
- xsv_debug
- xsv_line_number
- xsv_path
- xsv_version
- zipfile
- zipfile_cds
